relay-server/src/actors/project_buffer.rs

```diff
@@ -113,66 +159,237 @@ impl FromMessage<RemoveMany> for Buffer {
+                        })
+                        .collect::<Vec<_>>();
+
+                    // Since we have 3 variables we have to bind, we devide the SQLite limit by 3
+                    // here to prepare the chnunks which will be preparing the batch inserts.
+                    for chunk in envelopes.chunks(SQLITE_LIMIT_VARIABLE_NUMBER / 3) {
+                        query_builder.push_values(chunk, |mut b, v| match &v.1 {
+                            Ok(envelope_bytes) => {
+                                b.push_bind(v.0.own_key.to_string())
+                                    .push_bind(v.0.sampling_key.to_string())
+                                    .push_bind(envelope_bytes);
+                            }
+                            Err(err) => {
+                                relay_log::error!(
+                                    "failed to serialize the envelope: {}",
+                                    LogError(&err)
+                                )
+                            }
+                        });
+
+                        let query = query_builder.build();
+                        if let Err(err) = query.execute(&db).await {
+                            relay_log::error!(
+                                "failed to buffer envelopes to disk: {}",
+                                LogError(&err)
+                            )
+                        }
+                        // Reset the builder to initial state set by `QueryBuilder::new` function,
+                        // so it can be reused for another chunk.
+                        query_builder.reset();
+                    }
+                });
+            }
+        }
+        Ok(())
     }
 
     /// Handles the enqueueing messages into the internal buffer.
-    fn handle_enqueue(&mut self, message: Enqueue) {
-        self.buffer
-            .entry(message.key)
-            .or_default()
-            .push(message.value);
+    async fn handle_enqueue(&mut self, message: Enqueue) -> Result<(), BufferError> {
+        let Enqueue {
+            key,
+            value: managed_envelope,
+        } = message;
+
+        self.try_enqueue()?;
+
+        // save to the internal buffer
+        self.buffer.entry(key).or_default().push(managed_envelope);
+        self.count_mem_envelopes += 1;
+
+        Ok(())
     }
 
     /// Handles the dequeueing messages from the internal buffer.
     ///
     /// This method removes the envelopes from the buffer and stream them to the sender.
-    fn handle_dequeue(&mut self, message: DequeueMany) {
+    async fn handle_dequeue(&mut self, message: DequeueMany) -> Result<(), BufferError> {
         let DequeueMany { keys, sender } = message;
-        for key in keys {
-            for value in self.buffer.remove(&key).unwrap_or_default() {
+        for key in &keys {
+            for value in self.buffer.remove(key).unwrap_or_default() {
+                self.count_mem_envelopes -= 1;
```

Comment:
This does not really matter I think. But we could do it, yes.

Reply:
I think the order matters (best effort), but prioritizing one or another depends on what we want to achieve. But I don't know what we should prioritize: yielding from disk first vs memory will be slower, but it may also generate a situation where we are dealing with new envelopes while older ones are enqueued. I'd leave this as a follow-up to deal with when we have more information.